physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,684 @@
+		"server_stabilization_time":          ac.ServerStabilizationTime.String(),
+	})
+}
+
+// UnmarshalJSON parses the autopilot config JSON blob
+func (ac *AutopilotConfig) UnmarshalJSON(b []byte) error {
+	var data interface{}
+	err := json.Unmarshal(b, &data)
+	if err != nil {
+		return err
+	}
+
+	conf := data.(map[string]interface{})
+	if err = mapstructure.WeakDecode(conf, ac); err != nil {
+		return err
+	}
+	if ac.LastContactThreshold, err = parseutil.ParseDurationSecond(conf["last_contact_threshold"]); err != nil {
+		return err
+	}
+	if ac.DeadServerLastContactThreshold, err = parseutil.ParseDurationSecond(conf["dead_server_last_contact_threshold"]); err != nil {
+		return err
+	}
+	if ac.ServerStabilizationTime, err = parseutil.ParseDurationSecond(conf["server_stabilization_time"]); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// FollowerState represents the information about peer that the leader tracks.
+type FollowerState struct {
+	AppliedIndex  uint64
+	LastHeartbeat time.Time
+	LastTerm      uint64
+	IsDead        bool
+	NonVoter      bool
+}
+
+// FollowerStates holds information about all the followers in the raft cluster
+// tracked by the leader.
+type FollowerStates struct {
+	l         sync.RWMutex
+	followers map[string]FollowerState
+}
+
+// NewFollowerStates creates a new FollowerStates object
+func NewFollowerStates() *FollowerStates {
+	return &FollowerStates{
+		followers: make(map[string]FollowerState),
+	}
+}
+
+// markFollowerAsDead marks the node represented by the nodeID as dead. This
+// implies that Vault will indicate that the node "left" the cluster, the next
+// time autopilot asks for known servers. This function should be called with
+// the FollowerStates lock held.
+func (s *FollowerStates) markFollowerAsDead(nodeID string) {
+	state, ok := s.followers[nodeID]
+	if !ok {
+		return
+	}
+	s.followers[nodeID] = FollowerState{
+		LastHeartbeat: state.LastHeartbeat,
+		LastTerm:      state.LastTerm,
+		AppliedIndex:  state.AppliedIndex,
+		IsDead:        true,
+	}
+}
+
+// Update the peer information in the follower states
+func (s *FollowerStates) Update(nodeID string, appliedIndex uint64, term uint64, nonVoter bool) {
+	state := FollowerState{
+		AppliedIndex:  appliedIndex,
+		LastTerm:      term,
+		NonVoter:      nonVoter,
+		LastHeartbeat: time.Now(),
+	}
+
+	s.l.Lock()
+	s.followers[nodeID] = state
+	s.l.Unlock()
+}
+
+// Clear wipes all the information regarding peers in the follower states.
+func (s *FollowerStates) Clear() {
+	s.l.Lock()
+	for i := range s.followers {
+		delete(s.followers, i)
+	}
+	s.l.Unlock()
+}
+
+// Delete the entry of a peer represented by the nodeID from follower states.
+func (s *FollowerStates) Delete(nodeID string) {
+	s.l.Lock()
+	delete(s.followers, nodeID)
+	s.l.Unlock()
+}
+
+// Get retrieves information about a peer from the follower states.
+func (s *FollowerStates) Get(nodeID string) FollowerState {
+	s.l.RLock()
+	state := s.followers[nodeID]
```

Comment:
Good catch. 1) This wasn't used anywhere; so I ended up deleting the whole function. 2) If we want Get in future, it will be nice if we work with pointer to FollowerState instead. So updated the map to hold `*FollowerState`.